<!-- FroboScout launch file -->

<launch>
	<!-- Armadillo III parameters --> 
	<param name="diff_steer_wheel_distance" value="1.02"/>
	<param name="ticks_per_meter_left" value="740" />
	<param name="ticks_per_meter_right" value="740" />

	<node pkg="rosbag" name="rosbag_record" type="record" args="-a"/>

	<group ns="fmSensors">
		<node pkg="computer_load_monitor" name="load_monitor" type="computer_load_monitor_node" output="screen" >
			<param name="update_rate" value="5" />
		</node>

		<!-- extract data from messages transmitted by the VectorNav VN-100 IMU -->
		<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen"  respawn="true">
			<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
			<param name="imu_pub" value="/fmInformation/imu"/>
			<param name="frame_id" value="imu_link"/>
			<param name="use_enu" value="true"/>
			<param name="covariance_x" value="0.001"/>
			<param name="covariance_y" value="0.001"/>
			<param name="covariance_z" value="0.001"/>
		 </node>

		<!-- extract data from $GPGGA NMEA messages transmitted by the GPS -->
		<node pkg="nmea_to_tranmerc" name="nmea_to_tranmerc" type="nmea_to_tranmerc_node" output="screen" respawn="true">
			<param name="nmea_sub" value="/fmData/nmea_from_gps" />
			<param name="gpgga_tranmerc_pub" value="/fmInformation/gpgga_tranmerc" />
			<param name="transverse_mercator_a" value="6378137.0" />
			<param name="transverse_mercator_f" value="0.0033528106647474805" /> 
			<param name="transverse_mercator_false_easting" value="500000.0" />
			<param name="transverse_mercator_scale_factor" value="0.9996" />
			<param name="transverse_mercator_origin_latitude" value="0.0" />
			<param name="transverse_mercator_central_meridian" value="9.0" />
			<param name="transverse_mercator_false_northing" value="0.0" />
		</node>
	</group>

	<group ns="fmProcessors">
		<!-- calculate odometry based on encoders and IMU yaw axis orientation -->
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" respawn="true">
			<param name="enc_left_sub" value="/fmInformation/encoder_left" />
			<param name="enc_right_sub" value="/fmInformation/encoder_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odometry_reset" />
			<param name="odom_pub" value="/fmKnowledge/odometry" />
			<param name="odom_estimate_frame" value="/odom" />
			<param name="vehicle_frame" value="base_footprint" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="z" />
		</node>

		<node pkg="odometry_gnss_pose_estimator" name="pose_estimator" type="pose_estimator_node.py" output="screen" >
			<param name="odom_sub" value="/fmKnowledge/odometry" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gga_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="pose_pub" value="/fmKnowledge/pose" />
			<param name="pose_status_pub" value="/fmKnowledge/pose_status" />
			<param name="odometry_timeout" value="0.5" /> 
			<param name="gnss_timeout" value="2.0" /> 
			<param name="imu_timeout" value="0.5" />
			<param name="absolute_orientation_timeout" value="30" />
			<param name="odometry_distance_variance" value="0.000001" />
			<param name="odometry_angular_variance" value="0.000001" /> 
			<param name="gnss_angular_variance" value="0.0001" />
			<param name="solution_required" value="4" />
		</node>

		<!-- estimate robot 2d pose -->
		<!--node pkg="pose_2d_estimator" name="pose_estimator" type="pose_2d_estimator_node.py" output="screen" respawn="true">
			<param name="odom_sub" value="/fmKnowledge/odometry" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gga_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="pose_pub" value="/fmKnowledge/pose" />
			<param name="publish_relative_pose" type="boolean" value="false" />
			<param name="odometry_distance_variance" value="0.000001" /> 
			<param name="odometry_angular_variance" value="0.000035" />
		</node -->
	</group>

	<group ns="fmDecisionMakers">
		<node pkg="sdu_hudem_2014" name="mission" type="wiimote_mission_node_rtk_fixed.py" output="screen">
			<param name="wiimote_sub" value="/fmLib/joy" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="linear_velocity_default" value="0.2"/>
            <param name="linear_velocity_step" value="0.1"/>
            <param name="angular_velocity_default" value="0.2"/>
            <param name="angular_velocity_step" value="0.1"/>
            <param name="max_linear_velocity" value="0.4"/>
            <param name="max_angular_velocity" value="0.2"/>
            <param name="max_linear_acceleration" value="0.15"/>
            <param name="max_angular_acceleration" value="0.3"/>
            <param name="max_linear_deceleration" value="0.8"/>
            <param name="max_angular_deceleration" value="2.0"/>
		</node>

		<!--node pkg="sdu_hudem_2014" name="mission" type="keyboard_mission_node.py" output="screen">
			<param name="keyboard_sub" value="/fmHMI/keyboard" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="max_linear_velocity" value="0.8"/>
            <param name="max_angular_velocity" value="0.6"/>
            <param name="linear_velocity_step" value="0.1"/>
            <param name="angular_velocity_step" value="0.1"/>
		</node-->
	</group>

	<group ns="fmExecutors">
		<node pkg="sdu_hudem_2014" type="area_search_node.py" name="search" output="screen">
			<param name="automode_sub" value="/fmDecision/automode" />
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="joy_sub" value="/fmLib/joy" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="implement_pub" value="/fmCommand/implement" />
			<param name="status_pub" value="/fmInformation/wptnav_status" />
			<param name="pid_pub" value="/fmInformation/wptnav_pid" />
			<param name="status_publish_interval" value="1" />
			<param name="pid_publish_interval" value="1" />
			<param name="print_debug_information" value="false" />

            <param name="max_linear_velocity" value="0.5"/>
            <param name="max_angular_velocity" value="0.30"/>

            <param name="drive_kp" value="0.70"/>
            <param name="drive_ki" value="0.10"/>
            <param name="drive_kd" value="0.15"/>
            <param name="drive_feed_forward" value="0.00"/>
            <param name="drive_max_output" value="0.40"/>

            <param name="turn_kp" value="0.6"/>
            <param name="turn_ki" value="0.4"/>
            <param name="turn_kd" value="0.1"/>
            <param name="turn_feed_forward" value="0.0"/>
            <param name="turn_max_output" value="1.0"/>

			<param name="wpt_default_mode" value="MCTE" />
			<param name="wpt_default_tolerance" value="0.40" />
			<param name="wpt_default_drive_velocity" value="0.35" />
			<param name="wpt_default_turn_velocity" value="0.15" />
			<param name="wpt_default_wait_after_arrival" value="5.0" />
			<param name="wpt_default_implement_command" value="0.0" />

			<param name="turn_start_at_heading_err" value="45.0" />
			<param name="turn_stop_at_heading_err" value="2.0" />
			<param name="ramp_drive_velocity_at_distance" value="0.5" />
			<param name="ramp_min_drive_velocity" value="0.15" />
			<param name="ramp_turn_velocity_at_angle" value="35.0" />
			<param name="ramp_min_turn_velocity" value="0.10" />
			<param name="stop_navigating_at_distance" value="0.1" />

			<param name="casmo_width" value="0.8" />
			<param name="casmo_default_length" value="25" />

			<param name="wads_sub" value="/fmData/nmea_from_wads" />
			<param name="wads_turn_angle" value="0.6" />
			<param name="wads_turn_angular_speed_gain" value="5.0" />
			<param name="wads_penalty_distance" value="0.7" />
			<param name="wads_threshold" value="300" />
		</node>		
	</group>

	<group ns="fmControllers">
	</group>

	<group ns="fmLib">
		<node pkg="differential_ifk" type="differential_ifk_node" name="differential_ifk" output="screen">
			<param name="hl_subscriber_topic" value="/fmCommand/cmd_vel" />
			<param name="ll_publisher_topic_left" value="/fmSignals/cmd_vel_left" />
			<param name="ll_publisher_topic_right" value="/fmSignals/cmd_vel_right" />
			<param name="distance_center_to_wheel" value="1.10" />
			<param name="deadmanbutton_topic" value="/fmCommand/deadman" />
		</node>

		<node pkg="sdu_armadillo3" type="sdu_armadillo3_node" name="armadillo3_left" respawn="true">
			<param name="pid_topic" value="/fmInformation/pid_left" />
			<param name="cmd_vel_ch1_topic" value="/fmSignals/cmd_vel_left"/>
			<param name="serial_rx_topic" value="/fmData/left_belt_rx" />
			<param name="serial_tx_topic" value="/fmData/left_belt_tx" />
			<param name="deadman_topic" value="/fmCommand/deadman"/>
			<param name="encoder_ch1_topic" value="/fmInformation/encoder_left"/>
			<param name="propulsion_module_status_topic" value="/fmInformation/propulsion_module_status"/>
			<param name="propulsion_module_feedback_topic" value="/fmInformation/propulsion_module_feedback_left"/>
			<param name="power_ch1_topic" value="/fmData/power_left"/>
			<param name="status_topic" value="/fmData/status"/>
			<param name="ticks_per_meter" value="740" />
			<param name="mps_to_thrust" value="500"/>
			<param name="max_controller_command" value="800" />
			<param name="p_gain" value="1.0"/>
			<param name="i_gain" value="3.0"/>
			<param name="d_gain" value="0.1"/>
			<param name="feed_forward" value="1.5"/>
			<param name="i_max" value="0.1"/>
			<param name="closed_loop_operation" value="true"/>
			<param name="position_control" value="false"/>
			<param name="robot_max_acceleration" value="0.5"/>
			<param name="robot_max_jerk" value="0.5"/>
			<param name="brake_zeroband" value="0.2"/>  
			<param name="velocity_tolerance" value="0.05"/>  
		</node>

		<node pkg="sdu_armadillo3" type="sdu_armadillo3_node" name="armadillo3_right" respawn="true">
			<param name="pid_topic" value="/fmInformation/pid_right" />
			<param name="cmd_vel_ch1_topic" value="/fmSignals/cmd_vel_right"/>
			<param name="serial_rx_topic" value="/fmData/right_belt_rx" />
			<param name="serial_tx_topic" value="/fmData/right_belt_tx" />
			<param name="deadman_topic" value="/fmCommand/deadman"/>
			<param name="encoder_ch1_topic" value="/fmInformation/encoder_right"/>
			<param name="propulsion_module_status_topic" value="/fmInformation/propulsion_module_status"/>
			<param name="propulsion_module_feedback_topic" value="/fmInformation/propulsion_module_feedback_right"/>
			<param name="power_ch1_topic" value="/fmData/power_right"/>
			<param name="status_topic" value="/fmData/status"/>
			<param name="ticks_per_meter" value="740" />
			<param name="mps_to_thrust" value="500"/>
			<param name="max_controller_command" value="800" />
			<param name="p_gain" value="1.0"/>
			<param name="i_gain" value="3.0"/>
			<param name="d_gain" value="0.1"/>
			<param name="feed_forward" value="1.5"/>
			<param name="i_max" value="0.1"/>
			<param name="closed_loop_operation" value="true"/>  
			<param name="position_control" value="false"/>
			<param name="robot_max_acceleration" value="0.5"/>0
			<param name="robot_max_jerk" value="0.5"/>
			<param name="brake_zeroband" value="0.2"/>  
			<param name="velocity_tolerance" value="0.05"/>     
		</node>

		<!-- Wiimote interface -->
		<node pkg="wiimote" type="wiimote_node.py" name="wiimote" output="screen" respawn="true">
			<param name="fixed_wiimote_address" value="00:21:47:F0:56:EC"/> 
		</node>		

	 	<!--node pkg="keyboard_interface" type="keyboard_node" name="keyboard">
			<param name="keyboard_pub" value="/fmHMI/keyboard" />
		</node-->

		<node pkg="serial_string" type="serial_string_node" name="serial_comm_left" output="screen">
			<param name="publisher_topic" value="/fmData/left_belt_rx" />
			<param name="subscriber_topic" value="/fmData/left_belt_tx" />
			<param name="device" value="/dev/armadillo3_left" />
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="13" />
		</node>
	
		<node pkg="serial_string" type="serial_string_node" name="serial_comm_right" output="screen">
			<param name="publisher_topic" value="/fmData/right_belt_rx" />
			<param name="subscriber_topic" value="/fmData/right_belt_tx" />
			<param name="device" value="/dev/armadillo3_right" />
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="13" />
		</node>

		<!-- serial communication with the IMU -->
		<node pkg="serial_string" name="imu_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/imu_rx" />	 
			<param name="subscriber_topic" value="/fmData/imu_tx" />
			<param name="device" value="/dev/imu" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node> 	

		<!-- serial device and NMEA protocol handler for the RTK-GNSS -->
		<node pkg="serial_nmea" name="gps_nmea" type="serial_nmea_node" respawn="true" output="screen" >
			<param name="nmea_from_device_pub" value="/fmData/nmea_from_gps"/>
			<param name="nmea_to_device_sub" value="/fmData/nmea_to_gps"/>
			<param name="serial_device" value="/dev/gps"/>
			<param name="serial_baudrate" value="115200"/>
			<param name="use_nmea_checksum" value="true"/>
        </node>

		<!-- serial device and NMEA protocol handler for the WADS sensor -->
		<node pkg="serial_nmea" name="wads_nmea" type="serial_nmea_node" respawn="true" output="screen" >
			<param name="nmea_from_device_pub" value="/fmData/nmea_from_wads"/>
			<param name="nmea_to_device_sub" value="/fmData/nmea_to_wads"/>
			<param name="serial_device" value="/dev/wads" />	
			<param name="serial_baudrate" value="115200" />
			<param name="use_nmea_checksum" value="true"/>
		</node>

		<!--node pkg="real_time_analyze" name="rt_node" type="real_time_analyze_node.py" output="screen" >
			<param name="timing_pub" value="/fmInformation/rt_timing"/>
			<param name="device" value="/dev/rt_analyze" />	
        </node--> 	
	</group>
</launch>

